### PR TITLE
feature/#1576 - Additional samples with scalar invalidation from advanced polyline

### PR DESCRIPTION
**Pull request by mat8854**
See #1576 

Edit: Invalidation working now.

### CODE REVIEW

**Comment by monsieurtanuki** [comment]
@mat8854 For the record, I receive an email everytime you commit - in this case, 8 times so far.
Please send a specific comment when your code is ready for review.

**Comment by mat8854** [comment]
@monsieurtanuki It is ready now. I missunderstood the new `Container` usage and opened PR early. Now it all is working.

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        mMapping.add(60.f);
+        mMapping.add(100.f);
+        mMapping.add(100.f);
+        mContainer.add(80.f);
+        mContainer.add(60.f);
+        mContainer.add(100.f);
+        mContainer.add(100.f);
+
+        // update mapping with scalar end updated from 50 to 100
+        mMapping.init(0,100, 0,120);
```
Comment (monsieurtanuki) [inline OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:173]: Please don't use magic numbers: why 120, why 100?
* If 120 means something, create a dedicated constant like, `static final int MAX_HUE = 120;` (or double, I don't know)
* Same for 100: I should see `MAX_SCALAR`, or even a value computed from the scalar list (as the max value)

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        // call refresh to update line
+        mContainer.refresh();
+        // force a redraw (normally triggered when map is moved for example)
+        mMapView.invalidate();
+
+        // update UI
+        textViewCurrentLocation.setText("New scalar range from 0 to 100\nfor hue ranging from 0 to 120.");
```
Comment (monsieurtanuki) [inline OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:180]: Please use variables or constants in the string about the numbers: this way wa can understand the connection between the message and the code.

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        btnProceed.setText("Reset Polyline");
+
+    }
+
+    void centerToLine() {
```
Comment (monsieurtanuki) [inline OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:185]: make it `private`

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        scalars.add(50.0f);
+        scalars.add(25.0f);
+
+        // setup mapping objects
+        mMapping = new ColorMappingVariationHue(0,50, 0,120, 1.0f, 0.5f);
+        mContainer = new ColorMappingForScalarContainer(mMapping);
+
+        // add scalars to mapping and container
+        // you need to add the scalars to both lists
+        for (final float scalar : scalars) {
```
[the remark this answers is not in the document]
Comment (mat8854) [reply OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:129]: Okay, now I understood the design of the Container.

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        points.add(new GeoPoint(38.0, -11.0));
+        points.add(new GeoPoint(38.5, -11.5));
+        points.add(new GeoPoint(39.0, -11.0));
+        points.add(new GeoPoint(39.5, -11.5));
+
+        // create new polyline
+        mPolyline = new Polyline(mMapView, false, false);
+
+        // setup a black border
+        final Paint paintBorder = new Paint();
```
[the remark this answers is not in the document]
Comment (mat8854) [reply OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:101]: Okay.

File: OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java (hunk 1 above)
Comment (mat8854) [reply OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/drawing/ShowAdvancedPolylineStylesInvalidation.java:173]: I wanted to display "speed" ranging from red to green. That is 0 to 120:: https://i.stack.imgur.com/YOBFy.png

**Pull request merged**